### DISCUSSION TITLE
cocotb integration into a c++ and sv/uvm environment

### DISCUSSION THREAD

**Discussion by bqinTT** [cocotb/cocotb#4123] [Q&A] [answered]
Hey,

We started to integrate cocotb into our c++/sv/uvm environment and realized that upon cocotb finish execution it will call $finish which will override the uvm part which means all the uvm report summary is gone.
Now if you put cocotb test on a while(1) loop, the uvm code will call $finish. 

But cocotb will report simulator scheduler errors:

```
  2000.00ns ERROR    cocotb.scheduler                   Failing test at simulator request before test run completion: Simulator shut down prematurely
  2000.00ns ERROR    cocotb.regression                  Test error has lead to simulator shutting us down
  2000.00ns INFO     cocotb.regression                  **************************************************************************************
                                                        ** TEST                          STATUS  SIM TIME (ns)  REAL TIME (s)  RATIO (ns/s) **
                                                        **************************************************************************************
                                                        ** test_noc.noc_hello_world       FAIL        2000.00         168.15         11.89  **
                                                        **************************************************************************************
                                                        ** TESTS=1 PASS=0 FAIL=1 SKIP=0               2000.00         168.16         11.89  **
                                                        **************************************************************************************
```

Now I was wonder how can we get around this? 

Is there a way for cocotb not to call $finish? or cocotb has a hook manually call test completion tasks? We want to build a handshake way.

The flow order so far all the C++ test runs, and then sends a signal to uvm that is completed, now we are trying to link uvm with cocotb. 

Any ideas would be great! Only thing I can think of is ignore those message and false pass cocotb.

Thanks, Bo

**Comment by ktbarrett** [comment]
cocotb was designed to be "in charge" so there's no way to do this currently. You can probably hack something together by removing the calls to `simulator.stop_simulator` in regression.py.

cocotb is probably the wrong tool for the job for doing what you want, a more thorough solution would reuse the tools available to cocotb: GPI, PyGPI, scheduler, triggers, tasks, etc. but replace the regression manager and test system with something designed to act as a slave to the SV testbench. That work would be valuable, but hasn't been started yet.

**Comment by Dragon-Git** [comment, marked as answer]
> 
> Hey,
> 
> We started to integrate cocotb into our c++/sv/uvm environment and realized that upon cocotb finish execution it will call $finish which will override the uvm part which means all the uvm report summary is gone.
> Now if you put cocotb test on a while(1) loop, the uvm code will call $finish. 
> 
> But cocotb will report simulator scheduler errors:
> 
> ```
>   2000.00ns ERROR    cocotb.scheduler                   Failing test at simulator request before test run completion: Simulator shut down prematurely
>   2000.00ns ERROR    cocotb.regression                  Test error has lead to simulator shutting us down
>   2000.00ns INFO     cocotb.regression                  **************************************************************************************
>                                                         ** TEST                          STATUS  SIM TIME (ns)  REAL TIME (s)  RATIO (ns/s) **
>                                                         **************************************************************************************
>                                                         ** test_noc.noc_hello_world       FAIL        2000.00         168.15         11.89  **
>                                                         **************************************************************************************
>                                                         ** TESTS=1 PASS=0 FAIL=1 SKIP=0               2000.00         168.16         11.89  **
>                                                         **************************************************************************************
> ```
> 
> Now I was wonder how can we get around this? 
> 
> Is there a way for cocotb not to call $finish? or cocotb has a hook manually call test completion tasks? We want to build a handshake way.
> 
> The flow order so far all the C++ test runs, and then sends a signal to uvm that is completed, now we are trying to link uvm with cocotb. 
> 
> Any ideas would be great! Only thing I can think of is ignore those message and false pass cocotb.
> 
> Thanks, Bo
> 
 
You can try setting  set_finish_on_completion(0)  for the uvm_root. 
```systemverilog
uvm_root top;
top = uvm_root::get(); 
top.set_finish_on_completion(0);
```
This will prevent UVM from calling  $finish  when it finishes execution. This way, you can ensure that the UVM report summary is not overridden by Cocotb's  $finish  call.
 
Hope this helps! 😊

**Reply by bqinTT** [reply]
Nice! This works, I added some handshaking on my side. For uvm-1.2 it was finish_on_completion = 1. 

We can close this discussion.

**Comment by vincentatpku** [comment]
@Dragon-Git 
did you successfully maintain SV-based UVM components on cocotb framework?

**Reply by Dragon-Git** [reply]
I once ran a demo of a mix of cocotb and svuvm, but as mentioned in #1989, cocotb couldn't call sv functions via dpi, so I developed another library [svuvm](https://github.com/Dragon-Git/uvm_bridge), which is still experimental, using dpi as the top layer, and can now call some functions of svuvm in python, There are plans to integrate Cocotb in the future.

**Reply by vincentatpku** [reply]
my question is more about running cocotb test with SV-based uvm components. here is the case
we have a mature sv-based uvm testbench with driver/checker, etc.
can we integrate all these sv-based uvm env/comp in the cocotb framework? all these components would expecte to run in the background with time moving forwards. and the cocotb test case may just focus on enable/disable settings, etc.

Only when we solve the compatibility of cocotb & sv-based uvm components, the cocotb would become a more welcomed option for legacy.

**Reply by Dragon-Git** [reply]
Pls what problem are you having and I will try my best to tell me what I know?

**Reply by vincentatpku** [reply]
i have built up a cocotb framework for a smaill design block.
now i want to reuse some legacy checkers from sv-based testbench. these legacy class-checker is extended from uvm_components. 
how can i implement these legacy checker into cocotb framework? 
thanks for your quick reply.

**Reply by Dragon-Git** [reply]
You'll still need to use uvm_test to run the uvm_component-based checker, but uvm will call `$finish` by default, and you'll need to call `set_finish_on_completion` disable it.

**Reply by vincentatpku** [reply]
i suppose the items below are requiried.
1. include uvm library for cocotb compile/elab
2. add the legacy checkers into complie/elab list.

and when i run the cocotb_test, i would expect all these checkers are running in the background for checks. 

I feel the 1st challenge here is, whether i could introduce uvm library for cocotb compile/elab.

do you see the in-compatibility when including uvm library for cocotb comp/elab?

**Reply by Dragon-Git** [reply]
If you are not using DPI/VPI in your uvm environment, there should be no compatibility issues, and if you do, you may need to handle it carefully.

**Reply by vincentatpku** [reply]
could you share some code about how to introduce the uvm library into cocotb comp/elab?

**Reply by Dragon-Git** [reply]
I don't have the ready-made code on hand to provide, and I'm busy. Please be bold and do it, and if you have a specific question, I'll be happy to answer it for you.

**Comment by mballance** [comment]
If you're interested in calling SV functions and tasks from Python, you might be interested in the [PyHDL-IF](https://github.com/fvutils/pyhdl-if). This package can be used with cocotb or on its own, and enables calling SV tasks from Python coroutines.
